subsys/bluetooth/host/cs.c

```diff
@@ -22,6 +22,230 @@ LOG_MODULE_REGISTER(bt_cs);
+	return buf;
+}
+
+/** @brief Frees a reassembly buffer
+ *
+ * @note Takes the ownership of the pointer and sets it to NULL
+ *
+ * @param buf Double pointer to reassembly buffer
+ */
+static void free_reassembly_buf(struct net_buf **buf)
+{
+	if (!buf) {
+		LOG_ERR("NULL double pointer was passed when attempting to free reassembly buffer");
+		return;
+	}
+
+	if (!(*buf)) {
+		LOG_WRN("Attempted double free on reassembly buffer");
+		return;
+	}
+
+	struct reassembly_buf_meta_data *buf_meta_data =
+		(struct reassembly_buf_meta_data *)((*buf)->user_data);
+
+	LOG_DBG("De-allocating reassembly buffer for conn handle %d", buf_meta_data->conn_handle);
+	if (!sys_slist_find_and_remove(&reassembly_bufs, &(*buf)->node)) {
+		LOG_WRN("The buffer was not in the list");
+	}
+
+	net_buf_unref(*buf);
+	*buf = NULL;
+}
+
+/** @brief Gets the reassembly buffer identified by the connection handle
+ *
+ * @param conn_handle Connection handle
+ * @param allocate Allocates a new reassembly buffer if it's not allocated already
+ * @return struct net_buf* Reassembly buffer, NULL if it doesn't exist or failed when allocating new
+ */
+static struct net_buf *get_reassembly_buf(uint16_t conn_handle, bool allocate)
+{
+	sys_snode_t *node;
+
+	SYS_SLIST_FOR_EACH_NODE(&reassembly_bufs, node) {
+		struct net_buf *buf = CONTAINER_OF(node, struct net_buf, node);
+		struct reassembly_buf_meta_data *buf_meta_data =
+			(struct reassembly_buf_meta_data *)(buf->user_data);
+
+		if (buf_meta_data->conn_handle == conn_handle) {
+			return buf;
+		}
+	}
+
+	return allocate ? alloc_reassembly_buf(conn_handle) : NULL;
+}
+
+/** @brief Adds step data to a reassembly buffer
+ *
+ * @param reassembly_buf Reassembly buffer
+ * @param data Step data
+ * @param data_len Step data length
+ * @return true if successful, false if there is insufficient space
+ */
+static bool add_reassembly_data(struct net_buf *reassembly_buf, const uint8_t *data,
+				uint16_t data_len)
+{
+	if (data_len > net_buf_tailroom(reassembly_buf)) {
+		LOG_ERR("Not enough reassembly buffer space for subevent result");
+		return false;
+	}
+
+	net_buf_add_mem(reassembly_buf, data, data_len);
+	return true;
+}
+
+/** @brief Initializes a reassembly buffer from partial step data
+ *
+ * @note Upon first call, this function also registers the disconnection callback
+ *       to ensure any dangling reassembly buffer is freed
+ *
+ * @param conn_handle Connection handle
+ * @param steps Step data
+ * @param step_data_len Step data length
+ * @return struct net_buf* Pointer to reassembly buffer, NULL if fails to allocate or insert data
+ */
+static struct net_buf *start_reassembly(uint16_t conn_handle, const uint8_t *steps,
+					uint16_t step_data_len)
+{
+	static bool disconn_cb_registered;
+	struct net_buf *reassembly_buf = get_reassembly_buf(conn_handle, true);
+
+	if (!reassembly_buf) {
+		LOG_ERR("No buffer allocated for the result reassembly");
+		return NULL;
+	}
+
+	if (!disconn_cb_registered) {
```

Comment:
Might be simpler just to do 
```
BT_CONN_CB_DEFINE(cs_conn_callbacks) = {
	.disconnected = clear_on_disconnect,
};
```
Instead somewhere in the file?

Reply:
Of course!